src/server/chromium/videoRecorder.ts

```diff
@@ -84,58 +85,43 @@ export class VideoRecorder {
     this._gracefullyClose = gracefullyClose;
   }
 
-  async writeFrame(frame: Buffer, timestamp: number) {
+  writeFrame(frame: Buffer, timestamp: number) {
     assert(this._process);
-    if (!this._isRunning())
+    if (this._isStopped)
       return;
     this._progress.log(`writing frame ` + timestamp);
-    if (this._lastFrameBuffer)
-      this._lastWritePromise = this._flushLastFrame(timestamp - this._lastFrameTimestamp).catch(e => this._progress.log('Error while writing frame: ' + e));
+
+    if (this._lastFrameBuffer) {
+      const durationSec = timestamp - this._lastFrameTimestamp;
+      const repeatCount = Math.max(1, Math.round(fps * durationSec));
+      for (let i = 0; i < repeatCount; ++i)
+        this._frameQueue.push(this._lastFrameBuffer);
+      this._lastWritePromise = this._lastWritePromise.then(() => this._sendFrames());
+    }
+
     this._lastFrameBuffer = frame;
     this._lastFrameTimestamp = timestamp;
-    this._lastWriteTimestamp = Date.now();
+    this._lastWriteTimestamp = monotonicTime();
   }
 
-  private async _flushLastFrame(durationSec: number): Promise<void> {
-    assert(this._process);
-    const frame = this._lastFrameBuffer;
-    if (!frame)
-      return;
-    const previousWrites = this._lastWritePromise;
-    let finishedWriting: () => void;
-    const writePromise = new Promise<void>(fulfill => finishedWriting = fulfill);
-    const repeatCount = Math.max(1, Math.round(fps * durationSec));
-    this._progress.log(`flushing ${repeatCount} frame(s)`);
-    await previousWrites;
-    for (let i = 0; i < repeatCount; i++) {
-      const callFinish = i === (repeatCount - 1);
-      this._process.stdin.write(frame, (error: Error | null | undefined) => {
-        if (error)
-          this._progress.log(`ffmpeg failed to write: ${error}`);
-        if (callFinish)
-          finishedWriting();
-      });
-    }
-    return writePromise;
+  private async _sendFrames() {
+    while (this._frameQueue.length)
+      await this._sendFrame(this._frameQueue.shift()!);
+  }
+
+  private async _sendFrame(frame: Buffer) {
+    return new Promise(f => this._process!.stdin.write(frame, f)).then(error => {
+      if (error)
+        this._progress.log(`ffmpeg failed to write: ${error}`);
+    });
   }
 
   async stop() {
-    if (!this._gracefullyClose)
+    if (this._isStopped)
       return;
-
-    if (this._lastWriteTimestamp) {
-      const durationSec = (Date.now() - this._lastWriteTimestamp) / 1000;
-      if (!this._lastWritePromise || durationSec > 1 / fps)
-        this._flushLastFrame(durationSec).catch(e => this._progress.log('Error while writing frame: ' + e));
-    }
-
-    const close = this._gracefullyClose;
-    this._gracefullyClose = null;
+    this.writeFrame(Buffer.from([]), this._lastFrameTimestamp + (monotonicTime() - this._lastWriteTimestamp) / 1000);
```

Comment:
nit: wrap in  `if (this._lastWriteTimestamp) {` for the case when there are now frames.

Reply:
We are waiting for a frame now unconditionally.